lib/Doctrine/DBAL/Driver/PDOSqlsrv/Connection.php

```diff
@@ -25,13 +25,13 @@ public function lastInsertId($name = null)
         $stmt = $this->prepare('SELECT CONVERT(VARCHAR(MAX), current_value) FROM sys.sequences WHERE name = ?');
         $stmt->execute([$name]);
 
-        return $stmt->fetchColumn();
+        return (string) $stmt->fetchColumn();
```

Comment:
Indeed!

I have a problem regarding exceptions, though: I have documented `Driver\Connection::lastInsertId()` to throw a `Driver\DriverException` on error, but the only current implementations of this interface are specific to other drivers: `MysqliException`, `OCI8Exception` etc. I cannot use `Driver\PDOException`, which even though it implements the interface, expects a native `\PDOException` in its constructor, and I can only pass an error message string here.

One way to make things right, is to create a `PdoSqlSrvException` class implementing `Driver\DriverException`, but I'm wondering: why is `DriverException` an interface, and why are there several driver-specific implementations?

I cannot think of a use case where someone would want to catch an `OCI8Exception` specifically, for example. You're using a `Driver\Connection` **interface**, that tells you that methods throw a `Driver\DriverException`. Why would you want to catch a subclass of that?

I guess the primary reason why `Driver\DriverException` was an interface, is that `Driver\PDOException` needed at some point to extend `PDOException`, and at the same time be an instanceof `Driver\PDOException`. This is not the case anymore: `PDOConnection` does not extend `\PDO` anymore, so I see no reason to keep `Driver\PDOException` extending `\PDOException`.

As such, I would suggest to:

- drop all driver-specific exception classes, including `Driver\PDOException`
- make `Driver\DriverException` an instantiable class instead
- make all drivers throw `DriverException`s directly

What do you think?

I you're happy with the concept but think this is too much for this PR, I can create a `PDOSqlSrvException` class for now, and after this PR is merged, create another PR to deal with exceptions.

Reply:
> I cannot use `Driver\PDOException`, which even though it implements the interface, expects a native `\PDOException` in its constructor, and I can only pass an error message string here.

I'm experiencing the same problem in https://github.com/doctrine/dbal/pull/3154.

> One way to make things right, is to create a `PdoSqlSrvException` class implementing `Driver\DriverException`, but I'm wondering: why is `DriverException` an interface, and why are there several driver-specific implementations?
> 
> I cannot think of a use case where someone would want to catch an `OCI8Exception` specifically, for example. You're using a `Driver\Connection` **interface**, that tells you that methods throw a `Driver\DriverException`. Why would you want to catch a subclass of that?

You wouldn't want to catch a specific implementation of an interface but as long as it's an interface, implementations can be independent on each other. E.g. have methods with the same name but different signatures like `fromError(array $error)` and `fromError(PDOException $e)`.

> As such, I would suggest to:
> 
> * drop all driver-specific exception classes, including `Driver\PDOException`
> * make `Driver\DriverException` an instantiable class instead
> * make all drivers throw `DriverException`s directly
> 
> What do you think?

I think we need to leave the interface for now. There is an ongoing work https://github.com/doctrine/dbal/pull/3131 around reorganizing exceptions, so I'd like us to refrain from introducing any major changes.

If it's only the case for PDO (I assume all other driver-specific classes can be used), can we use something like:
```php
throw new class ($message) extends \Doctrine\DBAL\Driver\AbstractDriverException {};
```
It should satisfy the interface and let us move forward. It will be cleaned up later once have a better understanding of the approach.